### ISSUE TITLE
Memory leaks if used from a DLL

### ISSUE THREAD

**Issue by 00P** [Microsoft/cpprestsdk#74]
I all,
I started to write a DLL project that uses the nuget cpprestsdk.V140.winapp version 2.7.0.
The DLL is a x64 MFC DLL type.
Memory leaks appear just if have a global object that looks like:
     #include <cpprest/json.h>
     web::json::value MyVal;
and I exit application by normal quit().
I load and release the DLL that use cpprest140d_2_7.dll by AfxLoadLibrary() and AfxFreeLibrary() respectively.

Sorry if I posted this task here and mybe it is inappropriate for this place.

Some idea/help?
Thank you in advantage.

**Comment by ras0219-msft** [comment]
Do you have any more information about the leaks? Where were they allocated / what size are they / etc?

**Comment by 00P** [comment]
Hi, I can see if have saved some test project with the apparent problem to send you.


Best regards,

Giorgio


Inviato da Outlook Mobile. Sì, funziona con Gmail.






On Wed, Mar 2, 2016 at 3:45 PM -0800, "Robert Schumacher" <notifications@github.com> wrote:





Do you have any more information about the leaks? Where were they allocated / what size are they / etc?

---

**Comment by PremanandPatil** [comment]
Hi All,

Currently I am trying to use basic Proxy Authentication on Windows. It is
working fine for all kind of request except multi-part request.

When I am trying to send multi-part request I am getting "407-Proxy
Authentication Required".

Can I get help to fix this issue ?

-Prem.

**Comment by 00P** [comment]
Hi,

I found 2 very simple test project (one exe that call a DLL) and they show mem leaks.

If you have a ftp I can upload both the project for you.

 

Note that I can see mem leaks if the caller (the exe) is compiled from VS2013 and not if VS2015.

 

However the problem is simple to replicate, I think that if you use similar code you can replicate easily.


The code embedded in an OnClick button event of an exec (for example).



       int i=10;//Increasing DLL Load/Unload cycle Memory Leacks will grow.

       while(i--){

              dllh=AfxLoadLibrary(LR"RAW(Test_casablanca_in_QTDLL.dll)RAW");

              if(dllh){

                     (ExecTest)GetProcAddress(dllh,"ExecTest")();

                     AfxFreeLibrary(dllh);

              }else

                     ::MessageBoxA(nullptr,"DLL non caricata.","no",MB_OK);

              Sleep(100);

       }

 

 

 

The ExecTest function into Test_casablanca_in_QTDLL.dll  DLL :



 

 

 

#ifdef __cplusplus

extern "C"{

       __declspec(dllexport) bool ExecTest();

}

#endif

 

 

 

bool ExecTest(){

       web::json::value Body;

       return true;

}

 

 

 

 

 

 

The results:

 

 

 

Detected memory leaks!

Dumping objects ->

{35174} normal block at 0x00000226606EF570, 24 bytes long.

Data: <        `d a    > 01 00 00 00 CD CD CD CD 60 64 A8 61 FA 7F 00 00 

{35173} normal block at 0x00000226606EF510, 32 bytes long.

Data: <p n`&           > 70 F5 6E 60 26 02 00 00 00 00 00 00 00 00 00 00 

{35172} normal block at 0x00000226606EF270, 32 bytes long.

Data: <          n`&   > 01 00 00 00 CD CD CD CD 10 F5 6E 60 26 02 00 00 

{31320} normal block at 0x00000226606EF630, 24 bytes long.

Data: <        `d a    > 01 00 00 00 CD CD CD CD 60 64 A8 61 FA 7F 00 00 

{31319} normal block at 0x00000226606EF690, 32 bytes long.

Data: <0 n`&           > 30 F6 6E 60 26 02 00 00 00 00 00 00 00 00 00 00 

{31318} normal block at 0x00000226606EF330, 32 bytes long.

Data: <          n`&   > 01 00 00 00 CD CD CD CD 90 F6 6E 60 26 02 00 00 

{27466} normal block at 0x00000226606EF150, 24 bytes long.

Data: <        `d a    > 01 00 00 00 CD CD CD CD 60 64 A8 61 FA 7F 00 00 

{27465} normal block at 0x00000226606EEF70, 32 bytes long.

Data: <P n`&           > 50 F1 6E 60 26 02 00 00 00 00 00 00 00 00 00 00 

{27464} normal block at 0x00000226606EF1B0, 32 bytes long.

Data: <        p n`&   > 01 00 00 00 CD CD CD CD 70 EF 6E 60 26 02 00 00 

{23612} normal block at 0x00000226606C6850, 24 bytes long.

Data: <        `d a    > 01 00 00 00 CD CD CD CD 60 64 A8 61 FA 7F 00 00 

{23611} normal block at 0x00000226606EA720, 32 bytes long.

Data: <Phl`&           > 50 68 6C 60 26 02 00 00 00 00 00 00 00 00 00 00 

{23610} normal block at 0x00000226606EB380, 32 bytes long.

Data: <          n`&   > 01 00 00 00 CD CD CD CD 20 A7 6E 60 26 02 00 00 

{19758} normal block at 0x00000226606EAB40, 24 bytes long.

Data: <        `d a    > 01 00 00 00 CD CD CD CD 60 64 A8 61 FA 7F 00 00 

{19757} normal block at 0x00000226606EA8A0, 32 bytes long.

Data: <@ n`&           > 40 AB 6E 60 26 02 00 00 00 00 00 00 00 00 00 00 

{19756} normal block at 0x00000226606EA840, 32 bytes long.

Data: <          n`&   > 01 00 00 00 CD CD CD CD A0 A8 6E 60 26 02 00 00 

{15904} normal block at 0x00000226606EB560, 24 bytes long.

Data: <        `d a    > 01 00 00 00 CD CD CD CD 60 64 A8 61 FA 7F 00 00 

{15903} normal block at 0x00000226606EAE40, 32 bytes long.

Data: <` n`&           > 60 B5 6E 60 26 02 00 00 00 00 00 00 00 00 00 00 

{15902} normal block at 0x00000226606EB200, 32 bytes long.

Data: <        @ n`&   > 01 00 00 00 CD CD CD CD 40 AE 6E 60 26 02 00 00 

{12050} normal block at 0x00000226606EAF00, 24 bytes long.

Data: <        `d a    > 01 00 00 00 CD CD CD CD 60 64 A8 61 FA 7F 00 00 

{12049} normal block at 0x00000226606EADE0, 32 bytes long.

Data: <  n`&           > 00 AF 6E 60 26 02 00 00 00 00 00 00 00 00 00 00 

{12048} normal block at 0x00000226606EAD20, 32 bytes long.

Data: <          n`&   > 01 00 00 00 CD CD CD CD E0 AD 6E 60 26 02 00 00 

{8196} normal block at 0x000002266075FDB0, 24 bytes long.

Data: <        `d a    > 01 00 00 00 CD CD CD CD 60 64 A8 61 FA 7F 00 00 

{8195} normal block at 0x00000226607602F0, 32 bytes long.

Data: <  u`&           > B0 FD 75 60 26 02 00 00 00 00 00 00 00 00 00 00 

{8194} normal block at 0x0000022660760470, 32 bytes long.

Data: <          v`&   > 01 00 00 00 CD CD CD CD F0 02 76 60 26 02 00 00 

{4342} normal block at 0x00000226606EA900, 24 bytes long.

Data: <        `d a    > 01 00 00 00 CD CD CD CD 60 64 A8 61 FA 7F 00 00 

{4341} normal block at 0x00000226606EB500, 32 bytes long.

Data: <  n`&           > 00 A9 6E 60 26 02 00 00 00 00 00 00 00 00 00 00 

{4340} normal block at 0x00000226606EA9C0, 32 bytes long.

Data: <          n`&   > 01 00 00 00 CD CD CD CD 00 B5 6E 60 26 02 00 00 

{487} normal block at 0x00000226606F21B0, 24 bytes long.

Data: <        `d a    > 01 00 00 00 CD CD CD CD 60 64 A8 61 FA 7F 00 00 

{486} normal block at 0x00000226606F2030, 32 bytes long.

Data: < !o`&           > B0 21 6F 60 26 02 00 00 00 00 00 00 00 00 00 00 

{485} normal block at 0x00000226606F2510, 32 bytes long.

Data: <        0 o`&   > 01 00 00 00 CD CD CD CD 30 20 6F 60 26 02 00 00 

Object dump complete.

The program '[17980] testCasablanca.exe' has exited with code 0 (0x0).

 

 

 

 

I hope it is a false flag or help you to fix.

 

 

Sorry for my English.

 

Cheers,

Giorgio

**Comment by bakerrl2** [comment]
Did you write an Extension DLL or a Regular DLL?

**Comment by 00P** [comment]
Regular dll. Project under vs 2015 but with platform tool on VS 2013.

Dll project builded from QT visual studio plugin.

If you have an ftp i will upload all for you.


Cheers,

Giorgio


Inviato da Outlook Mobile. Sì, funziona con Gmail.






On Tue, Mar 15, 2016 at 11:39 AM -0700, "bakerrl2" <notifications@github.com> wrote:





Did you write an Extension DLL or a Regular DLL?

---

**Comment by BlackSoldierB** [comment]
Same issue as #67?

**Comment by bakerrl2** [comment]
No. I am not having the same problem has #67.  In my case I am trying to use the pplx tasks and http_client in a Regular DLL to make request to a web service.  When I call the wait function on the task the code just hangs and never processes the request or returns.

**Comment by bakerrl2** [comment]
See issue #93 for more of a description of my problem.
